### ISSUE TITLE
Consider kepub

### ISSUE THREAD

**Issue by gbnewby** [gutenbergtools/ebookmaker#169]
The Koboized epubs have some advantages, and Kobo has enough popularity to warrant some of the special treatment we give to Kindle.

Here's a converter with some details on the differences: https://pgaskin.net/kepubify/

It looks like Calibre can also convert to kepub.

**Comment by eshellman** [comment]
I would base the code for this on https://github.com/standardebooks/tools/blob/master/se/vendor/kobo_touch_extended/kobo.py to avoid lots of overhead, as we can start with a prebuilt tree.

My main concerns would be validation and testing.

**Comment by eshellman** [comment]
Roger and I have discovered that all we need to do is to rename the file from `.epub` to `.kepub.epub` and our epub3 files invoke the "good" rendering engine!

**Comment by tangledhelix** [comment]
@eshellman Out of curiosity:

If the solution were that easy, I assume we could have the web site present `.kepub.epub` download links (which point to the `-epub3.epub` file content) and solve the problem without spending more processing capacity or disk storage.

Since that hasn't happened, I'm curious if the missing element is doing more testing toward proving the theory that a filename change is enough. If we need such testing, I'll volunteer to do some.

I've identified a number of improvements that do seem to work (in a small sampling so far) by merely renaming the epub3 file to `.kepub.epub`:

- epub3 default font size is way too large compared to the device baseline, kepub fixes it
- illustrated drop caps are hidden in both renderings, but epub3 still offsets the margins the drop-cap would have consumed, throwing off layout badly
- The cover image thumbnail display is improved
- Page numbers in the right margin are correctly suppressed in kepub
- Footnotes in epub3 are basically untappable; kepub fixes the tap zones to work properly

All of the above is observed with the rename of epub3, **or** by using kepubify; so far I've found no differences.

I did find a couple of downgrades between epub3 and kepub. However, those are the same whether I rename the file or run kepubify, so again, there's no apparent difference.

1. page-breaks seem to be less distinct than in epub3 (no `<hr>` or actual page-break)
2. Some gesperrt spacing I've been working on got worse... but it's an edge case.

Let me know if more testing is the missing element. I recently bought a Kobo Libra and would be happy to try to move this forward, especially if the solution on the server would be as simple as offering an additional filename.

**Comment by eshellman** [comment]
@tangledhelix 
Dan, You are correct, the "missing element" is lack of testing support.
The only complication with just changing the download name is that (I think) it's best done in the .htaccess file rather than in ebookmaker. @gbnewby can help with that. The upside to that is that we can test everything without touching ebookmaker or autocat3, and fully implement with an easy change to autocat3!

Thanks for your offer of support, these are encouraging results.

**Comment by gbnewby** [comment]
@tangledhelix and Eric, would you please confirm what's being suggested?

What I'm seeing in this thread is that I should modify
https://dev.gutenberg.org's .htaccess file so that downloads of files
ending in "-epub3.epub" are exposed and named by the browser as
".kepub.epub" (such as, 'pg11119.kepub.epub").

Is that right?

My experience with my Kobo is that the main annoyance with PG's epub files
is that page numbering is incorrect. I've also noticed problems with
dropcaps and footnotes.

Thanks in advance for your confirmation of the experiment that is being
suggested.
  ~ Greg

**Comment by eshellman** [comment]
to be specific, you need to add a disposition http header to the server response
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition￼
Content-Disposition - HTTP | MDN
developer.mozilla.org

**Comment by eshellman** [comment]
what exactly did you rename the files to?

**Comment by tangledhelix** [comment]
> what exactly did you rename the files to?

It doesn't seem to matter so long as it ends in `.kepub.epub`, i.e. `foo.kepub.epub` - I assume replacing `foo` with the book number would be the preference in production.

**Comment by tangledhelix** [comment]
Or I should say - the book number seems like it would be preferred on gutenberg.org. (I know ebookmaker is used elsewhere too.)

**Comment by gbnewby** [comment]
Ok, I set up one book with this.

Please visit: https://dev.gutenberg.org/ebooks/65869

And select EPUB3 (E-readers incl. Send-to-Kindle)
<https://dev.gutenberg.org/ebooks/65869.epub3.images>

It should download as pg65869-images-3.kepub.epub

**Comment by tangledhelix** [comment]
Thanks, @gbnewby, I grabbed that book. I think I can also grab random epub3 files from the main site and rename them locally before putting them on the device. My plan was to fetch a bunch of those and start to compile notes about any rendering differences I find between the filenames and then share the notes.

It'll probably be a while before I'm done with that, as I have to find the time to do it.

**Comment by gbnewby** [comment]
Thanks for this update. We're quite interested in the results!

**Comment by eshellman** [comment]
@gbnewby I see that you've implemented that with a symlink. Are you thinking that's a better approach than adding the Content-Disposition header with .htaccess?

**Comment by gbnewby** [comment]
No, it's not better. I could not figure out the .htaccess syntax to rename
the download. If you provide the syntax, I can implement it. (Or, you can
experiment - edit ~/www/dev/html/.acl.gutenwebdev).

The symlink obviously only works for that one book, and is also visible in
PROD, so it's not the long-term solution if Dan's experiments support the
idea of simply renaming the download for Kobo compatibility.

**Comment by eshellman** [comment]
I'm definitely not an htaccess header expert but it would be something like the examples in 
https://help.dreamhost.com/hc/en-us/articles/215747598-Setting-headers-with-an-htaccess-file

using something like 
`Header set Content-Disposition "pgXXXX-images.kepub.epub"`

**Comment by eshellman** [comment]
I looked at an htaccess file. I don't have the current one. But...

after
`RewriteRule ^ebooks/([0-9]+)\.epub\.images$   /cache/epub/$1/pg$1-images.epub [L,R]`

I would have (following the existing naming pattern)

`
SetEnvIf Request_URI "^ebooks/([0-9]+)\.kepub\.images" PGID="$1"
Header set Content-Disposition "expr=pg%{PGID}-images.kepub.epub" env=PGID
RewriteRule ^ebooks/([0-9]+)\.kepub\.images$   /cache/epub/$1/pg$1-images.epub [L]
`
note, you don't want external redirect here.

**Comment by gbnewby** [comment]
Yes, this is the type of thing we'll need when autocat3 presents the
.kepub.epub link on the landing page.

The Request_URI from the client browser will have .kepub.images (or
whatever autocat3 uses) and we'll then need to (a) map to an actual
filename, like *-3.images.epub) and (b) rename the download with the
.kepub.epub suffix.

I don't think this .htaccess syntax will be useful until we change autocat3
to offer the kepub download link (unless Dan creates his own download link,
rather than using the landing page). Tell me if I'm misunderstanding
something and this can help even with the current autocat3 setup and
cache/epub/ contents.

**Comment by eshellman** [comment]
@tangledhelix would you be ok with pasting in urls like https://dev.gutenberg.org/ebooks/13982.kepub.images for testing?

@gbnewby The added header tells the client to rename the file, we don't do that server-side. That's why we redirect server-side, not client-side.

**Comment by tangledhelix** [comment]
@eshellman I have no issue with it, but to test this idea I can rename the epub3 files after I download them, so there's no server config needed merely to test the idea that the filename is the only difference from what we already have. What I'm researching is only this question: is renaming an epub3 file just as good as running through a converter (kepubify) and storing a second file?

I'm happy to help test the server redirect stuff--but it's not a dependency for what I'm doing for the time being.

**Comment by eshellman** [comment]
kepubify looks like a winner. Implementation should take less time than I've already wasted on trying to get apache to do the content disposition header

**Comment by tangledhelix** [comment]
Maybe so. I only tried this on one book so probably a bit more testing is warranted. But I also wonder if we should look at Calibre again. It was mentioned earlier but not talked about recently. The kepubify repos seems to not get much attention, and I wonder if Calibre would more well-supported into the future. I'll try to play with it when I get time, and see if the converter can be automated. From what I've read, it can convert epub to kepub and also edit kepub files.

**Comment by tangledhelix** [comment]
Kobo seems to sometimes handle text wrapping by just splitting the word wherever it runs out of space and starting again on the next line. Not only not using a hyphen, but not splitting at the place a hyphen ought to be. For example, I saw this in a sidenote.

```
Superintenden
t Registrar.
```

If we use kepubify, I think we definitely want to use `--hyphenate`, which changes the above, at least with the SR file I'm testing, my font sizes, etc, to:

```
Superin-
tendent Regis-
trar.
```

**Comment by tangledhelix** [comment]
Quick update here - elsewhere discussion has moved away from epubify and toward Calibre. I posted above that Calibre conversion didn't work properly for me, but I have since come to think I converted incorrectly in that test.

I'm going to do some more testing with Calibre-converted books and report back here.

**Comment by eshellman** [comment]
The Calibre incorporate kepubify code?

**Comment by tangledhelix** [comment]
> The Calibre incorporate kepubify code?

As far as I can tell, no, there's no connection aside from some function names in Calibre also use the term "kepubify". The kepubify program we looked at before is a Go program; all the Calibre code involved in converting to kepub is Python.
